server/src/main/java/io/crate/replication/logical/action/TransportAlterPublicationAction.java

```diff
@@ -0,0 +1,194 @@
+                    throw new PublicationUnknownException(request.name);
+                } else {
+                    var publication = oldMetadata.publications().get(request.name);
+                    if (publication != null) {
+                        var newPublication = updatePublication(request, currentMetadata, publication);
+
+                        PublicationsMetadata newMetadata = PublicationsMetadata.newInstance(oldMetadata);
+                        newMetadata.publications().put(request.name, newPublication);
+                        assert !newMetadata.equals(oldMetadata) : "must not be equal to guarantee the cluster change action";
+                        mdBuilder.putCustom(PublicationsMetadata.TYPE, newMetadata);
+
+                        return ClusterState.builder(currentState).metadata(mdBuilder).build();
+                    } else {
+                        throw new PublicationUnknownException(request.name);
+                    }
+                }
+            }
+
+            @Override
+            public void clusterStateProcessed(String source, ClusterState oldState, ClusterState newState) {
+                listener.onResponse(new AcknowledgedResponse(true));
+            }
+
+            @Override
+            public void onFailure(String source, Exception e) {
+                if (LOGGER.isTraceEnabled()) {
+                    logger.trace("Error while trying to alter publication " + request.name, e);
+                }
+                listener.onFailure(e);
+            }
+        };
+
+        clusterService.submitStateUpdateTask("alter-publication", updateTask);
+    }
+
+    @VisibleForTesting
+    static Publication updatePublication(Request request, Metadata currentMetadata, Publication oldPublication) {
+        // Ensure tables exists
+        for (var relation : request.tables) {
+            if (currentMetadata.hasIndex(relation.indexNameOrAlias()) == false
+                && currentMetadata.templates().containsKey(PartitionName.templateName(relation.schema(), relation.name())) == false) {
+                throw new RelationUnknown(relation);
+            }
+        }
+
+        HashSet<RelationName> tables = new HashSet<>();
+        switch (request.operation) {
+            case SET -> tables = new HashSet<>(request.tables);
+            case ADD -> {
+                tables.addAll(oldPublication.tables());
+                tables.addAll(request.tables);
+            }
+            case DROP -> oldPublication.tables().stream()
+                .filter(relationName -> request.tables.contains(relationName) == false)
+                .forEach(tables::add);
+            default ->
+                throw new UnsupportedOperationException(
+                    "Alter publication operation '" + request.operation + "' is not supported"
+                );
+        }
+        return new Publication(oldPublication.owner(), oldPublication.isForAllTables(), new ArrayList<>(tables));
```

Comment:
I'm checking currently PG docs - not clear to me how they handle case when a publication created with `ALL TABLES`:  as a user I expect either:
1. it's superior to set/add/drop so ALTER PUB is ignored for a pub with `all tables = true`?
2. we apply set/add/drop but set all tables false

Basically, it might be confusing to have both list of tables and ALL TABLES flag simultaneously .

Reply:
This is already prevented, see https://github.com/crate/crate/blob/master/server/src/main/java/io/crate/replication/logical/analyze/LogicalReplicationAnalyzer.java#L129 and https://github.com/crate/crate/blob/master/server/src/test/java/io/crate/replication/logical/analyze/LogicalReplicationAnalyzerTest.java#L159.

Reply:
Sorry, missed it, looked up only changed files 😅